File: js/packages/app/src/__tests__/features/object-references/external/polywrap/object-references-example/src/polywrap/external/global-functions/testInvokeExternalStaticMethod.ts
import { concat, u32ToBuffer, IExternalWrapInstance, BaseTypeSerialization } from '@polywrap/reim-wrap-js';
import { WrapModule } from '../module/WrapModule';
import { WrapManifest } from '../../WrapManifest';
import { ExternalResource } from '../../dt/ExternalResource';


export function testInvokeExternalStaticMethod(
  arg: string,
): Promise<string> {
  return testInvokeExternalStaticMethodFromInstance(
    WrapModule.wrapInstance,
    arg,
  );
};

export const create = (instance: IExternalWrapInstance) => {
  return (
    arg: string,
  ): Promise<string> => {
    return testInvokeExternalStaticMethodFromInstance(
      instance, 
      arg,
    );
  };
};

export const testInvokeExternalStaticMethodFromInstance = async (
  instance: IExternalWrapInstance | null, 
  arg: string,
): Promise<string> => {
  if (instance == null) {
    throw new Error("connect() or import() must be called before using this module");
  }

  const args = new TestInvokeExternalStaticMethodArgs(
    arg,
  );

  const buffer = concat([
    u32ToBuffer(WrapManifest.External.GlobalFunction.TestInvokeExternalStaticMethod),
    TestInvokeExternalStaticMethodArgsWrapped.serialize(args),
  ]);

  const result = await instance.invokeResource(ExternalResource.InvokeGlobalFunction, buffer);

  
  return BaseTypeSerialization.deserialize<string>(result);
}

export class TestInvokeExternalStaticMethodArgsWrapped {
  constructor(
    public arg: string,
  ) {
  }

  static serialize(value: TestInvokeExternalStaticMethodArgs): Uint8Array {
    return new TextEncoder().encode(
      JSON.stringify(
        TestInvokeExternalStaticMethodArgsWrapped.mapToSerializable(value)
      )
    );
  }

  static mapToSerializable(value: TestInvokeExternalStaticMethodArgs): TestInvokeExternalStaticMethodArgsWrapped {
    return new TestInvokeExternalStaticMethodArgsWrapped(
            
      value.arg,
      
    );
  }
}

export class TestInvokeExternalStaticMethodArgs {
  constructor(
    public arg: string,
  ) {
  }
}

File: rust/app/src/polywrap/external_references/src/polywrap/external/global-functions/testInvokeExternalStaticMethod.rs
use reim_wrap::{ ExternalWrapModule };
use module::{ WrapManifest };
use dt::{ ExternalResource };
use external_module::{ WrapModule };
use wrapped::{ testInvokeExternalStaticMethodWrapped };


pub async fn testInvokeExternalStaticMethod(
  arg: string,
) -> string {
  testInvokeExternalStaticMethodFromInstance(
    WrapModule.wrapInstance,
    arg,
  )
};

pub fn create(instance: dyn ExternalWrapModule) {
  |
    arg: string,
  | {
    testInvokeExternalStaticMethodFromInstance(
      instance, 
      arg,
    )
  }
}

pub async fn testInvokeExternalStaticMethodFromInstance (
  instance: Option<ExternalWrapModule>, 
  arg: string,
) -> string {
  if instance.is_none() {
    panic!("connect() or import() must be called before using this module");
  }

  let args = TestInvokeExternalStaticMethodArgs.new(
    arg,
  );

  let buffer = [
    WrapManifest.External.GlobalFunction.TestInvokeExternalStaticMethod.to_be_bytes(),
    TestInvokeExternalStaticMethodArgsWrapped.serialize(args),
  ].concat();

  let result = instance.invokeResource(ExternalResource.InvokeGlobalFunction, buffer).await;

  
  BaseTypeSerialization.deserialize<string>(result)
}

struct TestInvokeExternalStaticMethodArgsWrapped {
    pub arg: string,
}
impl TestInvokeExternalStaticMethodArgsWrapped {
    pub fn new(
      arg: string,
    ) -> TestInvokeExternalStaticMethodArgsWrapped {
      TestInvokeExternalStaticMethodArgsWrapped {
        arg,
      }
    }

    pub fn serialize(value: TestInvokeExternalStaticMethodArgs) -> Vec<u8> {
        json!(
            TestInvokeExternalStaticMethodArgsWrapped.mapToSerializable(value)
        ).as_bytes()
    }

    pub fn mapToSerializable(value: TestInvokeExternalStaticMethodArgs) -> TestInvokeExternalStaticMethodArgsWrapped {
        TestInvokeExternalStaticMethodArgsWrapped.new(
                        
            value.arg,
            
        )
    }
}

struct TestInvokeExternalStaticMethodArgs {
    pub arg: string,
}

impl TestInvokeExternalStaticMethodArgs {
  pub fn new(
    arg: string,
  ) -> TestInvokeExternalStaticMethodArgs {
    TestInvokeExternalStaticMethodArgs.new (
      arg,
    )
  }
}
